src/disk_io_thread_pool.cpp

```diff
@@ -0,0 +1,207 @@
+				l.lock();
+			}
+			else
+				t.detach();
+		}
+		m_threads.clear();
+	}
+
+	void disk_io_thread_pool::thread_active()
+	{
+		--m_num_idle_threads;
+		TORRENT_ASSERT(m_num_idle_threads >= 0);
+
+		int current_min = m_min_idle_threads;
+		while (m_num_idle_threads < current_min
+			&& !m_min_idle_threads.compare_exchange_weak(current_min, m_num_idle_threads));
+	}
+
+	bool disk_io_thread_pool::try_thread_exit(std::thread::id id)
+	{
+		int to_exit = m_threads_to_exit;
+		while (to_exit > 0 &&
+			!m_threads_to_exit.compare_exchange_weak(to_exit, to_exit - 1));
+		if (to_exit > 0)
+		{
+			std::unique_lock<std::mutex> l(m_mutex);
+			if (!m_abort)
+			{
+				auto new_end = std::remove_if(m_threads.begin(), m_threads.end()
+					, [id, &l](std::thread& t)
+				{
+					if (t.get_id() == id)
+					{
+						t.detach();
+						return true;
+					}
+					return false;
+				});
+				TORRENT_ASSERT(new_end != m_threads.end());
+				m_threads.erase(new_end, m_threads.end());
+				if (m_threads.empty()) m_idle_timer.cancel();
+			}
+		}
+		return to_exit > 0;
+	}
+
+	std::thread::id disk_io_thread_pool::first_thread_id()
+	{
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (m_threads.empty()) return std::thread::id();
+		return m_threads.front().get_id();
+	}
+
+	void disk_io_thread_pool::job_queued(int queue_size)
+	{
+		// this check is not strictly necessary
+		// but do it to avoid aquiring the mutex in the trivial case
+		if (m_num_idle_threads >= queue_size) return;
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (m_abort) return;
+
+		// reduce the number of threads requested to stop if we're going to need
+		// them for these new jobs
+		int to_exit = m_threads_to_exit;
+		while (to_exit > (std::max)(0, m_num_idle_threads - queue_size) &&
+			!m_threads_to_exit.compare_exchange_weak(to_exit
+				, (std::max)(0, m_num_idle_threads - queue_size)));
+
+		// now start threads until we either have enough to service
+		// all queued jobs without blocking or hit the max
+		for (int i = m_num_idle_threads
+			; i < queue_size && int(m_threads.size()) < m_max_threads
+			; ++i)
+		{
+			// if this is the first thread started, start the reaper timer
+			if (m_threads.empty())
+			{
+				m_idle_timer.expires_from_now(seconds(reap_idle_threads_interval));
+				m_idle_timer.async_wait([this](error_code const& ec) { reap_idle_threads(ec); });
+			}
+
+			// this keeps the io_service::run() call blocked from returning.
+			// When shutting down, it's possible that the event queue is drained
+			// before the disk_io_thread has posted its last callback. When this
+			// happens, the io_service will have a pending callback from the
+			// disk_io_thread, but the event loop is not running. this means
+			// that the event is destructed after the disk_io_thread. If the
+			// event refers to a disk buffer it will try to free it, but the
+			// buffer pool won't exist anymore, and crash. This prevents that.
+			std::shared_ptr<io_service::work> work =
```

Comment:
At least as of boost 1.60 `io_service::work` is not movable, but that's because it defines a copy constructor, which makes me wonder why it was heap allocated in the first place.